src/lib/group.ts

```diff
@@ -86,33 +114,106 @@ class Group {
    */
   add(g: Group) {
     if (this.#isConstant() && g.#isConstant()) {
-      if (this.x.toBigInt() === 0n) {
+      // we additionally check if g + 0 = g, because adding zero to g just results in g (and vise versa)
+      if (this.x.toBigInt() === 0n || this.#isZero().toBoolean()) {
         return g;
-      } else if (g.x.toBigInt() === 0n) {
+      } else if (g.x.toBigInt() === 0n || g.#isZero().toBoolean()) {
         return this;
       } else {
         let g_proj = Pallas.add(this.#toProjective(), g.#toProjective());
         return Group.#fromProjective(g_proj);
       }
     } else {
-      let [, x, y] = Snarky.group.add(this.#toTuple(), g.#toTuple());
-      return new Group({ x, y });
+      const { x: x1, y: y1 } = this;
+      const { x: x2, y: y2 } = g;
+
+      let zero = new Field(0);
+
+      let same_x = Provable.witness(Field, () => x1.equals(x2).toField());
+
+      let inf = checkFinite
+        ? zero
+        : Provable.witness(Field, () =>
+            x1.equals(x2).and(y1.equals(y2).not()).toField()
+          );
+
+      let inf_z = Provable.witness(Field, () => {
+        if (y1.equals(y2).toBoolean()) return zero;
+        else if (x1.equals(x2).toBoolean()) return y2.sub(y1).inv();
+        else return zero;
+      });
+
+      let x21_inv = Provable.witness(Field, () => {
+        if (x1.equals(x2).toBoolean()) return zero;
+        else return x2.sub(x1).inv();
+      });
+
+      let s = Provable.witness(Field, () => {
+        if (x1.equals(x2).toBoolean()) {
+          let x1_squared = x1.square();
+          return x1_squared.add(x1_squared).add(x1_squared).div(y1.add(y1));
+        } else return y2.sub(y1).div(x2.sub(x1));
+      });
+
+      let x3 = Provable.witness(Field, () => {
+        return s.square().sub(x1.add(x2));
+      });
+
+      let y3 = Provable.witness(Field, () => {
+        return s.mul(x1.sub(x3)).sub(y1);
+      });
+
+      let [, x, y] = Snarky.group.ecadd(
+        Group.from(x1.seal(), y1.seal()).#toTuple(),
+        Group.from(x2.seal(), y2.seal()).#toTuple(),
+        Group.from(x3, y3).#toTuple(),
+        inf.value,
+        same_x.value,
+        s.value,
+        inf_z.value,
+        x21_inv.value
+      );
+
+      // similarly to the constant implementation, just that I couldn't figure out a more efficient way to zero for addition with zero
+      // and the implementation above (original OCaml implementation) returns something wild -> g + 0 != g where it should be g + 0 = g
+      let gIsZero = g.#isZero();
+      let thisIsZero = this.#isZero();
+
+      // if either one is the negation of the other, we just return the zero element since g + (-g) = 0 - but the OCaml implementation doesn't pick that up
+      let isNegation = g.neg().equals(this);
+
+      let isNewElement = gIsZero.or(thisIsZero).not().and(isNegation.not());
+
+      return Provable.switch(
```

Comment:
similar to the constant implementation; if one of the operands is `zero`, we just return the original element `g`. If no operand is `zero`, we return `g1 + g2`, we also have to handle negation `g + (-g) = 0` since the original implementation didn't do so (it returned a seemingly random element)

Reply:
I expect this currently to throw an error if `this = g = (0,0)`, because `Provable.switch` expects a boolean vector with only one `1` set (this isn't part of the constraints, though)

Reply:
`Provable.switch` adds up the inputs times the booleans. So, constraint-wise, it's all good, because it's fine to add `0 + 0` if you want `0`